staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go

```diff
@@ -0,0 +1,279 @@
+}
+
+// CacheEncode implements runtime.CacheableObject interface.
+// It serializes the object and writes the result to given io.Writer trying
+// to first use the already cached result and falls back to a given encode
+// function in case of cache miss.
+// It assumes that for a given identifier, the encode function always encodes
+// each input object into the same output format.
+func (o *cachingObject) CacheEncode(id runtime.Identifier, encode func(runtime.Object, io.Writer) error, w io.Writer) error {
+	result := func() *serializationResult {
+		// Fast-path for getting from cache.
+		serializations := o.serializations.Load().(serializationsCache)
+		if result, exists := serializations[id]; exists {
+			return result
+		}
+
+		// Slow-path (that may require insert).
+		o.lock.Lock()
+		defer o.lock.Unlock()
+
+		serializations = o.serializations.Load().(serializationsCache)
+		// Check if in the meantime it wasn't inserted.
+		if result, exists := serializations[id]; exists {
+			return result
+		}
+
+		// Insert an entry for <id>. This requires copy of existing map.
+		newSerializations := make(serializationsCache)
+		for k, v := range serializations {
+			newSerializations[k] = v
+		}
+		result := &serializationResult{}
+		newSerializations[id] = result
+		o.serializations.Store(newSerializations)
+		return result
+	}()
+	result.once.Do(func() {
+		buffer := bytes.NewBuffer(nil)
+		result.err = encode(o.GetObject(), buffer)
+		result.raw = buffer.Bytes()
+	})
+	// Once invoked, fields of serialization will not change.
+	if result.err != nil {
+		return result.err
+	}
+	_, err := w.Write(result.raw)
+	return err
+}
+
+// GetObject implements runtime.CacheableObject interface.
+func (o *cachingObject) GetObject() runtime.Object {
+	result := o.object.DeepCopyObject().(metaRuntimeInterface)
+	func() {
+		o.lock.Lock()
+		defer o.lock.Unlock()
+		if o.selfLink != nil {
+			result.SetSelfLink(*o.selfLink)
+		}
+	}()
+	return result
+}
+
+// GetObjectKind implements runtime.Object interface.
+func (o *cachingObject) GetObjectKind() schema.ObjectKind {
+	return o.object.GetObjectKind()
+}
+
+// DeepCopyObject implements runtime.Object interface.
+func (o *cachingObject) DeepCopyObject() runtime.Object {
+	// DeepCopyObject on cachingObject is not expected to be called anywhere.
+	// However, to be on the safe-side, we implement it, though given the
+	// cache is only an optimization we ignore copying it.
+	result := &cachingObject{}
+	result.object = o.object.DeepCopyObject().(metaRuntimeInterface)
+	result.serializations.Store(make(serializationsCache))
+	func() {
+		o.lock.Lock()
+		defer o.lock.Unlock()
+		if o.selfLink != nil {
+			selfLink := *o.selfLink
+			result.selfLink = &selfLink
+		}
+	}()
+	return result
+}
+
+// The following functions implement metav1.Object interface:
+// - getters simply delegate for underlying object
+// - SetSelfLink stores the value passed in the first call
+// - all other setters ignore the passed value.
```

Comment:
this is pretty confusing... I expected all getters/setters to pass through

Reply:
This is a bit tricky. We can let them pass through and this will work, because noone is really using that.
But if we would e.g. create an encoder that is trying to overwrite one of the metadata, this can clearly break it for everyone. If course this is possible to fix - that encode instead of using CacheEncode should be calling GetObject() and operate on it. But that's relying on others.

I personally think that ignoring the values is safer - passing it through is easy to do, but I'm not sure it's what we really want.

Reply:
spoke offline. will switch setters to pass through and invalidate the cache (probably with locking). we don't actually expect these to get called frequently, so this is just for completeness and to fulfill the objectmeta interface

Reply:
done